monitoring/check_collectd_mlab_test.py

```diff
@@ -0,0 +1,301 @@
+
+  def testunit_sock_sendcmd_RETURNS_successfully(self):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.recv.side_effect = list('1 default reply\n')
+
+    returned_value = check_collectd_mlab.sock_sendcmd(
+        mock_sock, 'GETVAL "whatever"')
+
+    self.assertEqual(returned_value, 1)
+    mock_sock.send.assert_called_with('GETVAL "whatever"\n')
+
+  def testunit_sock_sendcmd_WHEN_receive_bad_reply_RETURNS_zero(self):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.recv.side_effect = list('not-a-number junk\n')
+
+    returned_value = check_collectd_mlab.sock_sendcmd(
+        mock_sock, 'GETVAL "whatever"')
+
+    self.assertEqual(returned_value, 0)
+    mock_sock.send.assert_called_with('GETVAL "whatever"\n')
+
+  def testunit_sock_connect_WHEN_no_socket_RAISES_Error(self):
+    self.assertRaises(
+        check_collectd_mlab.SocketConnectionError,
+        check_collectd_mlab.sock_connect, 'no_socket_name')
+
+  def testunit_sock_readline_WHEN_socket_error_RAISES_Error(self):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.recv.side_effect = socket.error('fake error')
+
+    self.assertRaises(
+        check_collectd_mlab.SocketReadlineError,
+        check_collectd_mlab.sock_readline, mock_sock)
+
+
+class MLabCollectdAssertionTests(unittest.TestCase):
+
+  def setUp(self):
+    self._testdata_dir = os.path.join(
+        os.path.dirname(check_collectd_mlab.__file__), 'testdata')
+    check_collectd_mlab.COLLECTD_BIN = (
+        os.path.join(self._testdata_dir, 'fake_bin'))
+    check_collectd_mlab.COLLECTD_NAGIOS = (
+        os.path.join(self._testdata_dir, 'fake_bin'))
+    check_collectd_mlab.COLLECTD_PID = (
+        os.path.join(self._testdata_dir, 'fake_pid'))
+    check_collectd_mlab.COLLECTD_UNIXSOCK = (
+        os.path.join(self._testdata_dir, 'fake_socket'))
+    check_collectd_mlab.VSYSPATH_BACKEND = (
+        os.path.join(self._testdata_dir, 'fake_backend'))
+    check_collectd_mlab.VSYSPATH_SLICE = (
+        os.path.join(self._testdata_dir, 'fake_slice'))
+
+  @mock.patch('check_collectd_mlab.sock_connect')
+  def testunit_assert_collectd_responds_WHEN_sock_sendcmd_fails(
+      self, mock_sock_connect):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    mock_sock.send.side_effect = socket.error('fake socket error')
+    mock_sock_connect.return_value = mock_sock
+
+    self.assertRaises(
+        check_collectd_mlab.SocketSendCommandError,
+        check_collectd_mlab.assert_collectd_responds)
+    mock_sock_connect.assert_called_with(
+        check_collectd_mlab.COLLECTD_UNIXSOCK)
+
+  @mock.patch('check_collectd_mlab.sock_connect')
+  def testunit_assert_collectd_responds_WHEN_sock_readline_fails(
+      self, mock_sock_connect):
+    mock_sock = mock.Mock(spec_set=socket.socket)
+    # After sending a default reply, send EOF ('') prematurely.
+    mock_sock.recv.side_effect = list('1 default reply\n') + ['']
```

Comment:
Ok. I think the approach with `FakeSocketIO` is clearer than using a mock now. So, just a clarifying comment on unitest.Mock `.side_effect`.

Originally, I used `.side_effect` because `check_collectd_mlab.sock_readline` calls `recv(1)` which reads one character at a time from the socket. And, the unittest.Mock interface limits our choices to `.return_value` which returns the value "as-is" or `.side_effect` which returns/calls each element in turn. Setting `mock_sock.recv.side_effect = ('1 default reply\n', '')` won't work because `check_collectd_malb.sock_readline` expects one character at a time and on the first call to `recv` the mock will return the entire string `'1 default reply\n'`.

Ideally, I might redesign unittest.Mock to support four attributes, `return_value`, `return_values`, `side_effect`, and `side_effects`, for single or a sequence of plain return values, and single or a sequence of actual "side effects" like exceptions or lambdas. This would prevent the current over-loading on `.side_effect` and cover all common uses cases with less ambiguity.